gnomad_hail/utils/variant_qc.py

```diff
@@ -262,80 +275,481 @@ def quantiles_to_bin_boundaries(quantiles: List[int]) -> Dict:
+            bin_id: hl.cond(
+                bin_ht.bin_stats[bin_id].merged_bins.contains(bin_ht[bin_id]),
+                bin_ht[bin_id] + hl.int(hl.rand_unif(0, bin_ht.bin_stats[bin_id].merged_bins[bin_ht[bin_id]] + 1)),
+                bin_ht.bin_stats[bin_id].global_bin_indices[bin_ht[bin_id]]
             )
-            for rank_id in rank_expr
+            for bin_id in bin_expr
         }
     )
 
     if desc:
-        ht = ht.annotate(
-            **{rank_id: n_bins - ht[rank_id] for rank_id in rank_expr}
+        bin_ht = bin_ht.annotate(
+            **{bin_id: n_bins - bin_ht[bin_id] for bin_id in bin_expr}
+        )
+
+    # Because SNV and indel rows are mutually exclusive, re-combine them into a single bin.
+    # Update the global bin_stats struct to reflect the change in bin names in the table
+    if stratify_snv_indel:
+        bin_expr_no_snv = {bin_id.rsplit("_", 1)[0] for bin_id in bin_ht.bin_stats}
+        bin_ht = bin_ht.annotate_globals(
+            bin_stats=hl.struct(
+                **{
+                    bin_id: hl.struct(
+                        **{snv: bin_ht.bin_stats[f"{bin_id}_{snv}"] for snv in ["snv", "indel"]}
+                    )
+                    for bin_id in bin_expr_no_snv
+                }
+            )
+        )
+
+        bin_ht = bin_ht.transmute(
+            **{
+                bin_id: hl.cond(
+                    ht[bin_ht.key].snv,
+                    bin_ht[f'{bin_id}_snv'],
+                    bin_ht[f'{bin_id}_indel']
+                )
+                for bin_id in bin_expr_no_snv
+            }
+        )
+
+    return bin_ht
+
+
+def create_binned_ht(
+        ht: hl.Table,
+        n_bins: int = 100,
+        singleton: bool = True,
+        biallelic: bool = True,
+        adj: bool = True,
+        add_substrat: Optional[Dict[str, hl.expr.BooleanExpression]] = None
+) -> hl.Table:
+    """
+    Annotates table with a bin, where variants are binned based on score into `n_bins` equally-sized bins.
+    Note that the following fields should be present:
+    - score
+    - ac - expected that this is the adj filtered allele count
+    - ac_raw - expected that this is the raw allele count before adj filtering
+
+    Computes bin numbers stratified by SNV / Indels and with the following optional sub bins
+    - singletons
+    - biallelics
+    - biallelic singletons
+    - adj
+    - adj biallelics
+    - adj singletons
+    - adj biallelic singletons
+
+    :param Table ht: Input table
+    :param int n_bins: Number of bins to bin into
+    :param bool singleton: Should bins be stratified by singletons
+    :param bool biallelic: Should bins be stratified by bi-alleleic variants
+    :param bool adj: Should bins be stratified by adj filtering
+    :param dictionary of boolean expr keyed by str add_substrat: Any additional stratifications for adding bins
+    :return table with bin number for each variant
```

Comment:
`:return` needs to have a colon after it (`:return:`) to be interpreted correctly.